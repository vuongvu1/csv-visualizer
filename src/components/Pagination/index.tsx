import React, { useCallback } from "react";
import SC from "./styles";

export enum PageSize {
  SMALL = 10,
  MEDIUM = 15,
  LARGE = 20,
}

interface Props {
  total: number;
  current: number;
  onChangePage?: (
    event: React.MouseEvent<HTMLButtonElement, MouseEvent>,
    page: number
  ) => void;
  pageSize?: PageSize;
  onChangePageSize?: (page: number) => void;
}

// TODO: Improve this component to support page total less than 3
const Pagination: React.FC<Props> = ({
  total,
  current,
  onChangePage,
  pageSize,
  onChangePageSize,
}) => {
  const isFirst = current === 1;
  const isLast = current === total;

  /**
   * `firstPageNum`, `secondPageNum` and `lastPageNum` are 3 values shown in the
   * pagination, the logic here is a bit tricky when the total pages is 2
   */
  const firstPageNum = isFirst
    ? current
    : isLast && total !== 2
    ? current - 2
    : current - 1;
  const secondPageNum = isFirst
    ? 2
    : isLast && total !== 2
    ? total - 1
    : current;
  const lastPageNum = isFirst ? 3 : isLast ? total : current + 1;

  const handleClickOnPage = useCallback(
    (nextPage: number) => (
      event: React.MouseEvent<HTMLButtonElement, MouseEvent>
    ) => onChangePage?.(event, nextPage),
    [onChangePage]
  );

  return (
    <SC.Container>
      <div>
        <SC.Button
          data-testid="firstBtnId"
          isFunctionKey
          disabled={isFirst}
          onClick={handleClickOnPage(1)}
        >
          First
        </SC.Button>
        <SC.Button
          isFunctionKey
          disabled={isFirst}
          onClick={handleClickOnPage(current - 1)}
        >
          Prev
        </SC.Button>
        <SC.Button
          active={current === firstPageNum}
          onClick={handleClickOnPage(firstPageNum)}
        >
          {firstPageNum}
        </SC.Button>
        {total > 1 && (
          <SC.Button
            active={current === secondPageNum}
            onClick={handleClickOnPage(secondPageNum)}
          >
            {secondPageNum}
          </SC.Button>
        )}
        {total > 2 && (
          <SC.Button
            active={current === lastPageNum}
            onClick={handleClickOnPage(lastPageNum)}
          >
            {lastPageNum}
          </SC.Button>
        )}
        <SC.Button
          isFunctionKey
          disabled={isLast}
          onClick={handleClickOnPage(current + 1)}
        >
          Next
        </SC.Button>
        <SC.Button
          isFunctionKey
          disabled={isLast}
          onClick={handleClickOnPage(total)}
        >
          Last
        </SC.Button>
      </div>
      <div>
        <strong>Page size</strong>
        {[PageSize.SMALL, PageSize.MEDIUM, PageSize.LARGE].map((size) => (
          <SC.Button
            data-testid={`pageSize${size}BtnId`}
            key={size}
            active={pageSize === size}
            isPageSizeKey
            onClick={() => onChangePageSize?.(size)}
          >
            {size}
          </SC.Button>
        ))}
      </div>
    </SC.Container>
  );
};

export default Pagination;
