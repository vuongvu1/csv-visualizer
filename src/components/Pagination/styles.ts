import styled, { css } from "styled-components";
import {
  neutralColor,
  neutralDarkColor,
  whiteColor,
  primaryColor,
} from "configs/colors";
import { Spacing } from "configs/styles";

interface ButtonProps {
  isFunctionKey?: boolean;
  isPageSizeKey?: boolean;
  disabled?: boolean;
  active?: boolean;
}

export default {
  Container: styled.div`
    display: flex;
    justify-content: space-between;
    align-items: center;
  `,
  Button: styled.button<ButtonProps>`
    padding: ${Spacing.small};
    border: 1px solid ${neutralDarkColor};
    background-color: ${whiteColor};
    min-width: 34px;
    height: 34px;

    ${(p) =>
      !p.disabled &&
      css`
        &:hover {
          background-color: ${neutralColor};
        }
      `}

    ${(p) =>
      p.isFunctionKey &&
      css`
        width: 46px;
      `}

    ${(p) =>
      p.isPageSizeKey &&
      css`
        margin-left: ${Spacing.small};
        height: 28px;
      `}

    ${(p) =>
      p.active &&
      css`
        font-weight: bold;
        color: ${whiteColor};
        background-color: ${primaryColor};

        &:hover {
          background-color: ${primaryColor};
        }
      `}
  `,
};
