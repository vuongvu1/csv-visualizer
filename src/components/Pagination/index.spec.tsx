import React from "react";
import { render, cleanup, fireEvent } from "@testing-library/react";
import Pagination, { PageSize } from "./index";

describe("Pagination", () => {
  afterEach(cleanup);

  test("should match snapshot", () => {
    const { container } = render(
      <Pagination total={10} current={2} pageSize={PageSize.MEDIUM} />
    );
    expect(container).toMatchSnapshot();
  });

  // TODO: Add similar tests for other page change buttons
  test("should trigger onChangePage handler", () => {
    const onChangePageMock = jest.fn();
    const { getByTestId } = render(<Pagination total={10} current={2} />);

    const firstBtn = getByTestId("firstBtnId");
    fireEvent.click(firstBtn);

    expect(onChangePageMock).toMatchSnapshot(1);
  });

  // TODO: Add similar tests for other page size change buttons
  test("should trigger onChangePageSize handler", () => {
    const onChangePageSizeMock = jest.fn();
    const { getByTestId } = render(
      <Pagination total={10} current={2} pageSize={PageSize.MEDIUM} />
    );

    const pageSizeSmallBtn = getByTestId(`pageSize${PageSize.SMALL}BtnId`);
    fireEvent.click(pageSizeSmallBtn);

    expect(onChangePageSizeMock).toMatchSnapshot(PageSize.SMALL);
  });
});
