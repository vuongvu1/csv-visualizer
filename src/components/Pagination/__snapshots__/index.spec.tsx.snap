// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pagination should match snapshot 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
}

.c1 {
  padding: 6px;
  border: 1px solid #707070;
  background-color: #FFFFFF;
  min-width: 34px;
  height: 34px;
  width: 46px;
}

.c1:hover {
  background-color: #EFEFEF;
}

.c2 {
  padding: 6px;
  border: 1px solid #707070;
  background-color: #FFFFFF;
  min-width: 34px;
  height: 34px;
}

.c2:hover {
  background-color: #EFEFEF;
}

.c3 {
  padding: 6px;
  border: 1px solid #707070;
  background-color: #FFFFFF;
  min-width: 34px;
  height: 34px;
  font-weight: bold;
  color: #FFFFFF;
  background-color: #3F7EF0;
}

.c3:hover {
  background-color: #EFEFEF;
}

.c3:hover {
  background-color: #3F7EF0;
}

.c4 {
  padding: 6px;
  border: 1px solid #707070;
  background-color: #FFFFFF;
  min-width: 34px;
  height: 34px;
  margin-left: 6px;
  height: 28px;
}

.c4:hover {
  background-color: #EFEFEF;
}

.c5 {
  padding: 6px;
  border: 1px solid #707070;
  background-color: #FFFFFF;
  min-width: 34px;
  height: 34px;
  margin-left: 6px;
  height: 28px;
  font-weight: bold;
  color: #FFFFFF;
  background-color: #3F7EF0;
}

.c5:hover {
  background-color: #EFEFEF;
}

.c5:hover {
  background-color: #3F7EF0;
}

@media (max-width:480px) {
  .c0 {
    -webkit-box-pack: center;
    -webkit-justify-content: center;
    -ms-flex-pack: center;
    justify-content: center;
  }

  .c0 > *:first-child {
    margin-bottom: 6px;
  }
}

<div>
  <div
    class="c0"
  >
    <div>
      <button
        class="c1"
        data-testid="firstBtnId"
      >
        First
      </button>
      <button
        class="c1"
      >
        Prev
      </button>
      <button
        class="c2"
      >
        1
      </button>
      <button
        class="c3"
      >
        2
      </button>
      <button
        class="c2"
      >
        3
      </button>
      <button
        class="c1"
      >
        Next
      </button>
      <button
        class="c1"
      >
        Last
      </button>
    </div>
    <div>
      <strong>
        Page size
      </strong>
      <button
        class="c4"
        data-testid="pageSize10BtnId"
      >
        10
      </button>
      <button
        class="c5"
        data-testid="pageSize15BtnId"
      >
        15
      </button>
      <button
        class="c4"
        data-testid="pageSize20BtnId"
      >
        20
      </button>
    </div>
  </div>
</div>
`;

exports[`Pagination should trigger onChangePage handler 1`] = `[MockFunction]`;

exports[`Pagination should trigger onChangePageSize handler 1`] = `[MockFunction]`;
